README.md
# Sequent Numbers Package

[![Latest Version on Packagist](https://img.shields.io/packagist/v/girover/sequent-numbers.svg?style=flat-square)](https://packagist.org/packages/girover/sequent-numbers)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/girover/sequent-numbers/run-tests?label=tests)](https://github.com/girover/sequent-numbers/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/girover/sequent-numbers/Check%20&%20fix%20styling?label=code%20style)](https://github.com/girover/sequent-numbers/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/girover/sequent-numbers.svg?style=flat-square)](https://packagist.org/packages/girover/sequent-numbers)

---
## Content

  - [Introduction](#introduction)
  - [Prerequisites](#prerequisites)
  - [Installation](#installation)
  - [Configuration](#configuration)
  - [Usage](#usage)
    - [Making Numbers](#making-numbers)
    - [Storing Numbers In Database](#storing-numbers-in-database)
  - [Testing](#testing)
  - [Changelog](#changelog)
  - [Contributing](#contributing)
  - [Security Vulnerabilities](#security-vulnerabilities)
  - [Credits](#credits)
  - [License](#license)


## Introduction
**girover/sequent-numbers** is a package for generating sequence of numbers.   
And possibility of storing them in database.
## Prerequisites
- Laravel 8+
- PHP 8+
- Mysql 5.7+
## Installation
You can add the package via **composer**:

```bash
composer require girover/sequent-numbers

```
Before installing the package you should configure your database.   

## Usage

 ### getting set of numbers
To make a set of numbers in memory, you can do this.

 ```php
    // InAController
    use \Girover\SequentNumbers\Numbers;

    $numbers = new Numbers;

    $numbers->from('00000')->to('99999')->get()
    // This will return Illuminate\Database\Eloquent\Collection

 ```

 To add some constraints on your numbers:

 ```php
    // InAController
    use \Girover\SequentNumbers\Numbers;

    $numbers = new Numbers;
    $numbers->from("0000")->to("9999");
    $numbers->query()->where('number', '>', '1000')->get();
    // This will return Illuminate\Database\Eloquent\Collection
    
    $numbers->query()->whereBetween('number', ["5555","7777"])->get();
    // This will return Illuminate\Database\Eloquent\Collection

 ```

 ### Storing Numbers in Database

 To store your created numbers in the database, you can do this:

 
 ```php
    // InAController
    use \Girover\SequentNumbers\Numbers;

    $numbers = new Numbers;
    $numbers->from("0000")->to("9999");

    $numbers->storeInTable('my_numbers_table');

 ```
 **Note**: If the name of table already exists in database, the table should has 
 a column called 'number' with varchar(255) type.

CHANGELOG.md
# Changelog

All notable changes to `sequent-numbers` will be documented in this file.

## 1.0.0 - 202X-XX-XX

- initial release


src/TableCreator.php
<?php

namespace Girover\SequentNumbers;

use Girover\SequentNumbers\Numbers;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Schema;


class TableCreator{

    /**
     * @var $table the name of the table to create.
     */
    protected $table;
    
    /**
     * @var $name the name of the table to create.
     */
    protected $name = 'numbers';
    
    /**
     * Set the name of the table
     * 
     * @param string $name
     * @return \Girover\SequentNumbers\TableMaker
     */
    public function name(string $name)
    {
        $this->name = $name;

        return $this;
    }

    /**
     * Creating the table in the database with the a specified name
     * 
     */
    public function create()
    {
        // No name for the table is provided
        if (is_null($this->name)) {
            return false;
        }

        // The table is already exists in database
        if (Schema::hasTable($this->name)) {
            throw new NumbersException("Table ( ".$this->name." ) already exists in database.", 1);
        }

        Schema::create($this->name, function($table){
            // $table->increments('id');
            $table->string('number')->unique();

            // $table->unique('number');
            $table->index('number');
        });

        return true;
    }

    public function fill()
    {
        // Table is not created yet
        if (!Schema::hasTable($this->name)) {
            return 0;
        }

        $numbersTable = new Numbers($this->game->first_card, $this->game->last_card);
        
        try {
            DB::insert("INSERT INTO `".$this->name."` (`number`) SELECT `number` FROM ".$numbersTable->table()->getValue());
            return 1;
        } catch (\Illuminate\Database\QueryException $th) {
            return -1;
        }

        return 1;
    }

    public function fillWithNumbers(string $from, string $to)
    {
        $numbers_table = new Numbers($from, $to);
        
        return $this->fill($numbers_table);
    }

    public function drop()
    {
        Schema::drop($this->name);
        return true;
    }

}


src/Numbers.php
<?php

/*
|--------------------------------------------------------------------------
| Virtual numbers maker
|--------------------------------------------------------------------------
|
| This class uses to create virtual table of set of numbers like: 00000->99999 | A000->Z999
| The table called `virtual_numbers_table`
| 
| Can be used like:
| $numbers = new Numbers; // OR
| $numbers = new Numbers("00000",'99999');
| $numbers = new Numbers("AA000",'ZZ999');
| $numbers->from("000")->to("999");
| $numbers->query()->get();
| $numbers->query()->whereBetween('10', '30')->get();
| $numbers->query()->where('number', '>', '40')->get();
| $numbers->form('10')->to('80')->query()->get();
|
| $numbers->storeInTable('numbers');
*/


namespace Girover\SequentNumbers;

use BadMethodCallException;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Schema;

class Numbers{

    /**
     * @var from_number :The collection of numbers should start with this.
     **/ 
    protected $from_number = 0;
    /**
     * @var from_number_as_array :Split the start number to an array 
     */
    protected $from_number_as_array = [];
    /** 
     * @var to_number :End the set of numbers with this number.
     */
    protected $to_number  = 9;
    /** 
     *@var to_number_as_array :Split the last number to an array
     */
    protected $to_number_as_array   = [];

    protected $digits     = ['0', '1', '2', '3', '4', '5', '6' ,'7' ,'8', '9'];
    protected $smallChars = ['a', 'b', 'c', 'd', 'e', 'f', 'g' ,'h' ,'i', 'j', 'k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'];
    protected $bigChars   = ['A', 'B', 'C', 'D', 'E', 'F', 'G' ,'H' ,'I', 'J', 'K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'];

    /**
     * Constructor
     * @param string $from :The first number in the set
     * @param string $to   :The last number in the set
     */
    public function __construct($from="0", $to="9") {

        $this->from_number = $from;
        $this->to_number = $to;
    }

    /**
     * Set the first number in the numbers' set
     *
     * @param string $from_number like: 10000, 'A0000', .....
     * @return Girover\SequentNumbers\VirtualNumbersMaker
     */
    public function fromNumber($from_number)
    {
        $this->from_number          = $from_number;
        $this->from_number_as_array = $this->splitFromNumber($from_number);
        
        return $this;
    }

    public function from($from_number)
    {
        return $this->fromNumber($from_number);
    }

    /**
     * Set the last number of the numbers' set
     *
     * @param string $to_number like: 99999, 'A9999', .....
     * @return Girover\SequentNumbers\VirtualNumbersMaker
     */
    public function toNumber($to_number)
    {
        $this->to_number        = $to_number;
        $this->to_number_as_array = $this->splitToNumber($to_number);
        return $this;
    }

    public function to($to_number)
    {
        return $this->toNumber($to_number);
    }

    /**
     * Set the first and last number in the numbers' set.
     *
     * @param string $from like: 0000, 'A000', .....
     * @param string $to like: 99999, 'Z9999', .....
     * @return Girover\SequentNumbers\VirtualNumbersMaker
     */
    public function between($from, $to)
    {
        $this->fromNumber($from);
        $this->toNumber($to);

        return $this;
    }

    /**
     * Convert the first number to array of chars. like: [0,0,0,0], ['A','0','0','0'], .....
     *
     * @param string the number to split to array
     */
    private function splitFromNumber($number)
    {
        $arr = str_split($number);
        for($i=1; $i<count($arr); $i++){
            if(in_array($arr[$i],$this->digits)){
                $arr[$i] = '0';
            }elseif(in_array($arr[$i],$this->bigChars)){
                $arr[$i] = 'A';
            }else{
                $arr[$i] = 'a';
            }
        }
        return $arr;
    }

    /**
     * Convert the last number to array of chars. like: [9,9,9,9], ['Z','9','9','9'], .....
     *
     * @param string the number to split to array
     */
    private function splitToNumber($number)
    {
        $arr = str_split($number);
        for($i=1; $i<count($arr); $i++){
            if(in_array($arr[$i],$this->digits)){
                $arr[$i] = '9';
            }elseif(in_array($arr[$i],$this->bigChars)){
                $arr[$i] = 'Z';
            }else{
                $arr[$i] = 'z';
            }
        }
        return $arr;
    }

    /**
     * Make the field name for the virtual numbers table
     * 
     * @return string
     */
    private function makeFieldName()
    {
        if(count($this->from_number_as_array) !== count($this->to_number_as_array)){
            return false;
        }

        $field = ' CONCAT(';
        for($i= count($this->from_number_as_array); $i>0; $i--){
            $field .='val'.$i.',';
        }

        $field = substr($field, 0, -1).') AS `number` ';

        return $field;
    }

    /**
     * Make virtual table with one field and value with one digit|char
     *
     * @param digit|char $form : like 1,2,"2","a","m"
     * @param digit|char $to   : like 1,2,"2","a","m"
     * @param digit|char $to   : like 1,2,"2","a","m"
     *
     * @return string like: (select "a" val5 unoin select "b" val5 ........union select "z" val5) table5
     */
    private function makeTableWithOneFieldOneChar($from, $to, $fieldNumber)
    {
        $table = '(';
        foreach (range($from, $to) as $number) {
            $table .= ' SELECT "'.$number.'" val'.$fieldNumber.' UNION ';
        }
        // remove last 'UNION' from the string
        $table = preg_replace('/\W\w+\s*(\W*)$/', '',$table);
        $table .= ') table'.$fieldNumber.' ';

        return $table;
    }

    /**
     * Cross join all 'TableWithOneFieldOneChar' to gether to make big table with a lot of numbers
     *
     * (select 1 val1 union select 2 val1) table1 CROSS JOIN (select 1 val2 union s...)table2 CROSS JOIN ....
     */
    private function crossJoinTables()
    {
        $val = 1;
        $crossJoin = '';

        for($i=count($this->from_number_as_array);$i>0;$i--){
            $crossJoin .= $this->makeTableWithOneFieldOneChar($this->from_number_as_array[$i-1],$this->to_number_as_array[$i-1],$val++);
            $crossJoin .= ' CROSS JOIN ';
        }
        // Remove the last 'CROSS JOIN' from the string
        return preg_replace( "/\s[a-z]+\s[a-z]+\s$/i", "", $crossJoin);
    }

    /**
     * Make the query that gets all numbers between 'from_number' and 'to_number'
     * 
     * @return string
     */
    public function sqlQuery()
    {
        $query  = 'SELECT '.$this->makeFieldName().' FROM ';
        $query .= $this->crossJoinTables();

        return $query;
    }

    /**
     * Table called `numbers` has one field called `number`
     * you can get the sql as string by: table()->getValue()
     *
     * @return \Illuminate\Database\Expression
     */
    public function table()
    {
        // return DB::raw(' (SELECT `virtual_numbers`.`number` FROM ('.$this->sqlQuery().') virtual_numbers WHERE `virtual_numbers`.`number` BETWEEN "'.$this->from_number.'" AND "'.$this->to_number.'" ORDER BY `virtual_numbers`.`number`) virtual_numbers_table ');
        return DB::raw(' (SELECT `virtual_numbers`.`number` FROM ('.$this->sqlQuery().') virtual_numbers WHERE `virtual_numbers`.`number` BETWEEN "'.$this->from_number.'" AND "'.$this->to_number.'" ORDER BY `virtual_numbers`.`number`) virtual_numbers_table ');
    }

    /**
     * Getting query builder for the numbers table
     * 
     * @return \Illuminate\Database\Query\Builder
     */
    public function query()
    {
        return DB::table($this->table());
    }


    public function storeInTable($table_name)
    {
        if (! Schema::hasTable($table_name)) {           
            (new TableCreator)->name($table_name)->create();
        }
        
        try {
            DB::insert("INSERT INTO `".$table_name."` (`number`) SELECT `number` FROM ".$this->table()->getValue());
            return true;
        } catch (\Throwable $th) {
            throw new NumbersException($th->getMessage(), 1);                    
        }
    }
}
